File: tools/testing/connection/nodejs-client/sonar.mjs
#!/usr/bin/env node
/**
 * Sonar API - Node.js Client Integration Test
 *
 * Authentication: Bearer Token
 * Base URL: https://sonarcloud.io or https://your-sonar.example.com (SonarQube)
 * Health Endpoint: GET /api/authentication/validate
 *
 * Uses internal packages:
 *   - @internal/fetch-proxy-dispatcher: Environment-aware proxy configuration
 *   - @internal/fetch-client: HTTP client with auth support
 *   - @internal/provider_api_getters: API key resolution
 *   - @internal/app-static-config-yaml: YAML configuration loading
 *
 * API Documentation:
 *   SonarCloud: https://sonarcloud.io/web_api
 *   SonarQube: https://docs.sonarqube.org/latest/extension-guide/web-api/
 */
import { fileURLToPath } from 'url';
import { dirname, resolve } from 'path';

// ============================================================================
// Project Setup
// ============================================================================
const __dirname = dirname(fileURLToPath(import.meta.url));
const PROJECT_ROOT = resolve(__dirname, '..', '..', '..', '..');

// Load static config
const { loadYamlConfig, config: staticConfig } = await import(
  resolve(PROJECT_ROOT, 'packages_mjs', 'app-static-config-yaml', 'src', 'index.mjs')
);
const configDir = resolve(PROJECT_ROOT, 'common', 'config');
await loadYamlConfig({ configDir });

// Import internal packages
const { getProxyDispatcher } = await import(
  resolve(PROJECT_ROOT, 'packages_mjs', 'fetch-proxy-dispatcher', 'src', 'index.mts')
);
const { createClient } = await import(
  resolve(PROJECT_ROOT, 'packages_mjs', 'fetch-client', 'src', 'index.mts')
);
const { SonarApiToken, ProviderHealthChecker } = await import(
  resolve(PROJECT_ROOT, 'packages_mjs', 'provider_api_getters', 'src', 'index.mjs')
);

// ============================================================================
// Configuration - Exposed for debugging
// ============================================================================
const provider = new SonarApiToken(staticConfig);
const apiKeyResult = provider.getApiKey();

const CONFIG = {
  // From provider_api_getters
  SONAR_TOKEN: apiKeyResult.apiKey,
  AUTH_TYPE: apiKeyResult.authType,

  // Base URL (from provider or override)
  BASE_URL: provider.getBaseUrl() || process.env.SONAR_BASE_URL || 'https://sonarcloud.io',

  // Dispatcher (from fetch-proxy-dispatcher)
  DISPATCHER: getProxyDispatcher(),

  // Proxy Configuration (set to override YAML/environment config)
  PROXY: process.env.HTTPS_PROXY || process.env.HTTP_PROXY || undefined,

  // SSL/TLS Configuration (runtime override, or undefined to use YAML config)
  SSL_VERIFY: false,  // Set to undefined to use YAML config

  // Debug
  DEBUG: !['false', '0'].includes((process.env.DEBUG || '').toLowerCase()),
};

// ============================================================================
// Client Factory
// ============================================================================
function createSonarClient() {
  return createClient({
    baseUrl: CONFIG.BASE_URL,
    dispatcher: CONFIG.DISPATCHER,
    auth: {
      type: 'bearer',
      rawApiKey: CONFIG.SONAR_TOKEN,
    },
    headers: {
      Accept: 'application/json',
    },
    proxy: CONFIG.PROXY,
    verify: CONFIG.SSL_VERIFY,
  });
}

// ============================================================================
// Health Check
// ============================================================================
async function healthCheck() {
  console.log('\n=== Sonar Health Check (ProviderHealthChecker) ===\n');

  const checker = new ProviderHealthChecker(staticConfig);
  const result = await checker.check('sonar');

  console.log(`Status: ${result.status}`);
  if (result.latency_ms) console.log(`Latency: ${result.latency_ms.toFixed(2)}ms`);
  if (result.message) console.log(`Message: ${result.message}`);
  if (result.error) console.log(`Error: ${result.error}`);

  return { success: result.status === 'connected', result };
}

// ============================================================================
// Sample API Calls using fetch-client
// ============================================================================
async function validateAuth() {
  console.log('\n=== Validate Authentication ===\n');

  const client = createSonarClient();

  try {
    const response = await client.get('/api/authentication/validate');

    console.log(`Status: ${response.status}`);
    console.log('Response:', JSON.stringify(response.data, null, 2));

    return { success: response.ok, data: response.data };
  } finally {
    await client.close();
  }
}

async function getSystemStatus() {
  console.log('\n=== Get System Status ===\n');

  const client = createSonarClient();

  try {
    const response = await client.get('/api/system/status');

    console.log(`Status: ${response.status}`);
    console.log('Response:', JSON.stringify(response.data, null, 2));

    return { success: response.ok, data: response.data };
  } finally {
    await client.close();
  }
}

async function listProjects(organization = null, pageSize = 10) {
  console.log(`\n=== List Projects (pageSize: ${pageSize}) ===\n`);

  const client = createSonarClient();
  const query = { ps: pageSize };
  if (organization) {
    query.organization = organization;
  }

  try {
    const response = await client.get('/api/projects/search', { query });

    console.log(`Status: ${response.status}`);
    if (response.ok && response.data) {
      const components = response.data.components || [];
      const paging = response.data.paging || {};
      console.log(`Total: ${paging.total || 0}`);
      components.slice(0, 10).forEach((project) => {
        console.log(`  - ${project.key}: ${project.name}`);
      });
    } else {
      console.log('Response:', JSON.stringify(response.data, null, 2));
    }

    return { success: response.ok, data: response.data };
  } finally {
    await client.close();
  }
}

async function getProjectStatus(projectKey) {
  console.log(`\n=== Get Project Status: ${projectKey} ===\n`);

  const client = createSonarClient();

  try {
    const response = await client.get('/api/qualitygates/project_status', {
      query: { projectKey },
    });

    console.log(`Status: ${response.status}`);
    if (response.ok && response.data) {
      const projectStatus = response.data.projectStatus || {};
      console.log(`Quality Gate: ${projectStatus.status}`);
      const conditions = projectStatus.conditions || [];
      conditions.slice(0, 5).forEach((condition) => {
        console.log(`  - ${condition.metricKey}: ${condition.status} (${condition.actualValue})`);
      });
    } else {
      console.log('Response:', JSON.stringify(response.data, null, 2));
    }

    return { success: response.ok, data: response.data };
  } finally {
    await client.close();
  }
}

async function getProjectMetrics(projectKey) {
  console.log(`\n=== Get Project Metrics: ${projectKey} ===\n`);

  const metricKeys = 'bugs,vulnerabilities,code_smells,coverage,duplicated_lines_density';

  const client = createSonarClient();

  try {
    const response = await client.get('/api/measures/component', {
      query: { component: projectKey, metricKeys },
    });

    console.log(`Status: ${response.status}`);
    if (response.ok && response.data) {
      const component = response.data.component || {};
      const measures = component.measures || [];
      console.log(`Project: ${component.name}`);
      measures.forEach((measure) => {
        console.log(`  - ${measure.metric}: ${measure.value}`);
      });
    } else {
      console.log('Response:', JSON.stringify(response.data, null, 2));
    }

    return { success: response.ok, data: response.data };
  } finally {
    await client.close();
  }
}

async function listIssues(projectKey = null, pageSize = 10) {
  console.log(`\n=== List Issues (pageSize: ${pageSize}) ===\n`);

  const client = createSonarClient();
  const query = { ps: pageSize, resolved: 'false' };
  if (projectKey) {
    query.componentKeys = projectKey;
  }

  try {
    const response = await client.get('/api/issues/search', { query });

    console.log(`Status: ${response.status}`);
    if (response.ok && response.data) {
      const issues = response.data.issues || [];
      const paging = response.data.paging || {};
      console.log(`Total: ${paging.total || 0}`);
      issues.slice(0, 10).forEach((issue) => {
        const severity = issue.severity || 'UNKNOWN';
        const type = issue.type || 'UNKNOWN';
        let message = issue.message || '';
        if (message.length > 50) message = message.slice(0, 50) + '...';
        console.log(`  - [${severity}] ${type}: ${message}`);
      });
    } else {
      console.log('Response:', JSON.stringify(response.data, null, 2));
    }

    return { success: response.ok, data: response.data };
  } finally {
    await client.close();
  }
}

// ============================================================================
// Run Tests
// ============================================================================
async function main() {
  console.log('Sonar API Connection Test (Node.js Client Integration)');
  console.log('='.repeat(55));
  console.log(`Base URL: ${CONFIG.BASE_URL}`);
  console.log(`Token: ${CONFIG.SONAR_TOKEN ? CONFIG.SONAR_TOKEN.slice(0, 10) + '...' : 'Not set'}`);
  console.log(`Auth Type: ${CONFIG.AUTH_TYPE}`);
  console.log(`Debug: ${CONFIG.DEBUG}`);

  await healthCheck();

  // Uncomment to run additional tests:
  // await validateAuth();
  // await getSystemStatus();
  // await listProjects('your-org');
  // await getProjectStatus('your-project-key');
  // await getProjectMetrics('your-project-key');
  // await listIssues();
}

main().catch(console.error);

File: tools/testing/connection/python-client/sonar.py
#!/usr/bin/env python3
"""
Sonar API - Python Client Integration Test

Authentication: Bearer Token
Base URL: https://sonarcloud.io or https://your-sonar.example.com (SonarQube)
Health Endpoint: GET /api/authentication/validate

Uses internal packages:
  - fetch_client: HTTP client with auth support
  - provider_api_getters: API key resolution
  - static_config: YAML configuration loading

API Documentation:
  SonarCloud: https://sonarcloud.io/web_api
  SonarQube: https://docs.sonarqube.org/latest/extension-guide/web-api/
"""
import asyncio
import os
import sys
from pathlib import Path

# ============================================================================
# Project Setup
# ============================================================================
PROJECT_ROOT = Path(__file__).resolve().parents[4]
sys.path.insert(0, str(PROJECT_ROOT / "packages_py" / "app_static_config_yaml" / "src"))
sys.path.insert(0, str(PROJECT_ROOT / "packages_py" / "provider_api_getters" / "src"))
sys.path.insert(0, str(PROJECT_ROOT / "packages_py" / "fetch_client" / "src"))

# Load static config
from static_config import load_yaml_config, config as static_config
config_dir = PROJECT_ROOT / "common" / "config"
load_yaml_config(config_dir=config_dir)

# Import internal packages
from fetch_client import create_client_with_dispatcher, AuthConfig
from provider_api_getters import SonarApiToken, ProviderHealthChecker

# ============================================================================
# Configuration - Exposed for debugging
# ============================================================================
provider = SonarApiToken(static_config)
api_key_result = provider.get_api_key()

CONFIG = {
    # From provider_api_getters
    "SONAR_TOKEN": api_key_result.api_key,
    "AUTH_TYPE": api_key_result.auth_type,

    # Base URL (from provider or override)
    "BASE_URL": provider.get_base_url() or os.getenv("SONAR_BASE_URL", "https://sonarcloud.io"),

    # SSL/TLS Configuration (runtime override, or None to use YAML config)
    "SSL_VERIFY": False,  # Set to None to use YAML config
    "CERT": os.getenv("CERT"),
    "CA_BUNDLE": os.getenv("CA_BUNDLE"),

    # Proxy Configuration (set to override YAML/environment config)
    "PROXY": os.getenv("HTTPS_PROXY") or os.getenv("HTTP_PROXY"),

    # Debug
    "DEBUG": os.getenv("DEBUG", "true").lower() not in ("false", "0"),
}


# ============================================================================
# Client Factory
# ============================================================================
def create_sonar_client():
    """Create a Sonar API client with standard configuration."""
    return create_client_with_dispatcher(
        base_url=CONFIG["BASE_URL"],
        auth=AuthConfig(
            type="bearer",
            raw_api_key=CONFIG["SONAR_TOKEN"],
        ),
        default_headers={"Accept": "application/json"},
        verify=CONFIG["SSL_VERIFY"],
        cert=CONFIG["CERT"],
        ca_bundle=CONFIG["CA_BUNDLE"],
        proxy=CONFIG["PROXY"],
    )


# ============================================================================
# Health Check
# ============================================================================
async def health_check():
    """Run health check using ProviderHealthChecker."""
    print("\n=== Sonar Health Check (ProviderHealthChecker) ===\n")

    checker = ProviderHealthChecker(static_config)
    result = await checker.check("sonar")

    print(f"Status: {result.status}")
    if result.latency_ms:
        print(f"Latency: {result.latency_ms:.2f}ms")
    if result.message:
        print(f"Message: {result.message}")
    if result.error:
        print(f"Error: {result.error}")

    return {"success": result.status == "connected", "result": result}


# ============================================================================
# Sample API Calls using fetch-client
# ============================================================================
async def validate_auth():
    """Validate authentication."""
    print("\n=== Validate Authentication ===\n")

    async with create_sonar_client() as client:
        response = await client.get("/api/authentication/validate")

        print(f"Status: {response.status}")
        print(f"Response: {response.data}")

        return {"success": response.ok, "data": response.data}


async def get_system_status():
    """Get system status (works without authentication)."""
    print("\n=== Get System Status ===\n")

    async with create_sonar_client() as client:
        response = await client.get("/api/system/status")

        print(f"Status: {response.status}")
        print(f"Response: {response.data}")

        return {"success": response.ok, "data": response.data}


async def list_projects(organization: str = None, page_size: int = 10):
    """List projects (SonarCloud requires organization parameter)."""
    print(f"\n=== List Projects (page_size: {page_size}) ===\n")

    params = {"ps": page_size}
    if organization:
        params["organization"] = organization

    async with create_sonar_client() as client:
        response = await client.get("/api/projects/search", params=params)

        print(f"Status: {response.status}")
        if response.ok and response.data:
            components = response.data.get("components", [])
            paging = response.data.get("paging", {})
            print(f"Total: {paging.get('total', 0)}")
            for project in components[:10]:
                print(f"  - {project.get('key')}: {project.get('name')}")
        else:
            print(f"Response: {response.data}")

        return {"success": response.ok, "data": response.data}


async def get_project_status(project_key: str):
    """Get project quality gate status."""
    print(f"\n=== Get Project Status: {project_key} ===\n")

    async with create_sonar_client() as client:
        response = await client.get(
            "/api/qualitygates/project_status",
            params={"projectKey": project_key}
        )

        print(f"Status: {response.status}")
        if response.ok and response.data:
            project_status = response.data.get("projectStatus", {})
            print(f"Quality Gate: {project_status.get('status')}")
            conditions = project_status.get("conditions", [])
            for condition in conditions[:5]:
                print(f"  - {condition.get('metricKey')}: {condition.get('status')} ({condition.get('actualValue')})")
        else:
            print(f"Response: {response.data}")

        return {"success": response.ok, "data": response.data}


async def get_project_metrics(project_key: str):
    """Get project metrics."""
    print(f"\n=== Get Project Metrics: {project_key} ===\n")

    metric_keys = "bugs,vulnerabilities,code_smells,coverage,duplicated_lines_density"

    async with create_sonar_client() as client:
        response = await client.get(
            "/api/measures/component",
            params={
                "component": project_key,
                "metricKeys": metric_keys
            }
        )

        print(f"Status: {response.status}")
        if response.ok and response.data:
            component = response.data.get("component", {})
            measures = component.get("measures", [])
            print(f"Project: {component.get('name')}")
            for measure in measures:
                print(f"  - {measure.get('metric')}: {measure.get('value')}")
        else:
            print(f"Response: {response.data}")

        return {"success": response.ok, "data": response.data}


async def list_issues(project_key: str = None, page_size: int = 10):
    """List issues (optionally filtered by project)."""
    print(f"\n=== List Issues (page_size: {page_size}) ===\n")

    params = {"ps": page_size, "resolved": "false"}
    if project_key:
        params["componentKeys"] = project_key

    async with create_sonar_client() as client:
        response = await client.get("/api/issues/search", params=params)

        print(f"Status: {response.status}")
        if response.ok and response.data:
            issues = response.data.get("issues", [])
            paging = response.data.get("paging", {})
            print(f"Total: {paging.get('total', 0)}")
            for issue in issues[:10]:
                severity = issue.get("severity", "UNKNOWN")
                issue_type = issue.get("type", "UNKNOWN")
                message = (issue.get("message", "")[:50] + "...") if len(issue.get("message", "")) > 50 else issue.get("message", "")
                print(f"  - [{severity}] {issue_type}: {message}")
        else:
            print(f"Response: {response.data}")

        return {"success": response.ok, "data": response.data}


# ============================================================================
# Run Tests
# ============================================================================
async def main():
    print("Sonar API Connection Test (Python Client Integration)")
    print("=" * 55)
    print(f"Base URL: {CONFIG['BASE_URL']}")
    print(f"Token: {CONFIG['SONAR_TOKEN'][:10] + '...' if CONFIG['SONAR_TOKEN'] else 'Not set'}")
    print(f"Auth Type: {CONFIG['AUTH_TYPE']}")
    print(f"Debug: {CONFIG['DEBUG']}")

    await health_check()

    # Uncomment to run additional tests:
    # await validate_auth()
    # await get_system_status()
    # await list_projects(organization="your-org")
    # await get_project_status("your-project-key")
    # await get_project_metrics("your-project-key")
    # await list_issues()


if __name__ == "__main__":
    asyncio.run(main())
